src/main/java/net/neoforged/neoforge/attachment/AttachmentType.java

```diff
@@ -198,17 +199,26 @@ public Builder<T> serialize(Codec<T> codec) {
          */
         public Builder<T> serialize(Codec<T> codec, Predicate<? super T> shouldSerialize) {
             Objects.requireNonNull(codec);
-            // TODO: better error handling
             return serialize(new IAttachmentSerializer<>() {
                 @Override
                 public T read(IAttachmentHolder holder, Tag tag, HolderLookup.Provider provider) {
-                    return codec.parse(provider.createSerializationContext(NbtOps.INSTANCE), tag).result().get();
+                    final DataResult<T> parsingResult = codec.parse(provider.createSerializationContext(NbtOps.INSTANCE), tag);
+                    return parsingResult.getOrThrow(msg -> buildException("read", msg));
                 }
 
                 @Nullable
                 @Override
                 public Tag write(T attachment, HolderLookup.Provider provider) {
-                    return shouldSerialize.test(attachment) ? codec.encodeStart(provider.createSerializationContext(NbtOps.INSTANCE), attachment).result().get() : null;
+                    if (!shouldSerialize.test(attachment)) {
+                        return null;
+                    }
+                    final DataResult<Tag> encodingResult = codec.encodeStart(provider.createSerializationContext(NbtOps.INSTANCE), attachment);
+                    return encodingResult.getOrThrow(msg -> buildException("write", msg));
+                }
+
+                private RuntimeException buildException(final String operation, final String error) {
+                    // TODO: Maybe find a way to let the user know *which* attachment has exploded?
```

Comment:
I think that's already printed when catching the exception. Can you check?

Reply:
The ID is in fact printed on [deserialization](https://github.com/neoforged/NeoForge/blob/83c257a12ae695966439be48e1917ae8318148b6/src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java#L157), but no such `catch`es exist for serialization. Such a `catch`-then-log behavior might be added there, if desired.

Though, this only covers a subset of the attachment functionalities. For example, if someone were to manually copy attachments via `IEntityExtension#copyAttachmentsFrom` for whatever reason, an error in serde wouldn't see an attachment ID associated to it due to `AttachmentInternals` not performing any such checks. Granted, it could be argued this is out-of-scope of the intended use case.

Reply:
Okay; would you want me to add a check in `serializeAttachments` too, just in case?

Reply:
Sure yeah why not